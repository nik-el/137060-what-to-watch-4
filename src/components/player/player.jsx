import React, {useRef, useState, useEffect} from 'react';
import PropTypes from 'prop-types';

export const Player = React.memo(function Player({forcePlay, previewSrc, posterSrc, autoPlay, muted}) {
  const videoEl = useRef(null);
  const [playStatus, setPlayingStatus] = useState(false);

  useEffect(() => {
    if (!forcePlay && !playStatus) {
      return;
    }

    if (forcePlay && !playStatus) {
      videoEl.current.play();
    } else if (playStatus && !forcePlay) {
      videoEl.current.pause();
    }

    setPlayingStatus(!videoEl.current.paused);
  }, [forcePlay]);


  return <video
    ref={videoEl}
    src={previewSrc}
    autoPlay={autoPlay}
    muted={muted}
    poster={posterSrc}
    className={`small-player` + (playStatus ? ` small-player_play` : ``)}
  />;

});

Player.propTypes = {
  // принудительное состояние воспроизведния/пауза извне
  forcePlay: PropTypes.bool,
  // ссылка на превью
  previewSrc: PropTypes.string.isRequired,
  // ссылка на постер
  posterSrc: PropTypes.string,
  // автоплей
  autoPlay: PropTypes.bool,
  // состояние звука
  muted: PropTypes.bool
};
