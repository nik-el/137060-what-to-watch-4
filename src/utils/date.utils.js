const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const getFormattedDate = (dateString) => {
  if (!dateString) {
    return ``;
  }

  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
