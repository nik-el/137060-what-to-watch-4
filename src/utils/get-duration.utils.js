const HOUR_MIN = 60;
const HOUR_SEC = 3600;

export const geFormatDuration = (time) => {
  const hours = Math.floor(time / HOUR_MIN);
  const min = time % HOUR_MIN;

  return `${hours}h ${min}m`;
};

export const geVideoDuration = (time) => {
  const secNum = parseInt(time, 10);
  let hours = Math.floor(secNum / HOUR_SEC);
  let minutes = Math.floor((secNum - (hours * HOUR_SEC)) / HOUR_MIN);
  let seconds = secNum - (hours * HOUR_SEC) - (minutes * HOUR_MIN);

  if (hours < 10) {
    hours = `0` + hours;
  }
  if (minutes < 10) {
    minutes = `0` + minutes;
  }
  if (seconds < 10) {
    seconds = `0` + seconds;
  }
  return hours + `:` + minutes + `:` + seconds;
};
