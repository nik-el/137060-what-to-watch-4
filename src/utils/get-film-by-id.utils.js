export const getFilmById = (films, id) => {
  if (!Array.isArray(films) || !id) {
    return null;
  }

  return films.find((item) => item.id === id);
};
