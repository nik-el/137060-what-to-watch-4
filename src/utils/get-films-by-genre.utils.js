export const getFilmsByGenre = (films, genre) => {
  if (!Array.isArray(films)) {
    return [];
  }
  if (!genre) {
    return films;
  }

  return films.filter((item) => item.genre === genre);
};
