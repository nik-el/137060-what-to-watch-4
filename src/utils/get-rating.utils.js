export const getRating = (rating) => {
  switch (true) {
    case (rating < 3):
      return `Bad`;
    case (rating >= 3 && rating < 5):
      return `Normal`;
    case (rating >= 5 && rating < 8):
      return `Good`;
    case (rating >= 8 && rating < 10):
      return `Very good`;
    case (rating === 10):
      return `Awesome`;
    default:
      return `No rating`;
  }
};
