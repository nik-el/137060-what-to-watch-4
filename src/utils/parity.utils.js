const getOddArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }
  return array.filter((item, index) => {
    return index % 2;
  });

};
const getEvenArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }

  return array.filter((item, index) => {
    return !(index % 2);
  });
};

export {
  getOddArray,
  getEvenArray
};
