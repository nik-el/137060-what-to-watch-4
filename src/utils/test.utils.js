const testFilm = {
  title: `Test title 1`,
  genre: `test genre 1`,
  releaseYear: `2000`,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testFilm);

const testFilms = [
  {
    title: `Test title 1`,
    genre: `test genre 1`,
    releaseYear: `2000`,
    thumbnail: `https://picsum.photos/id/1`,
    bgImage: `https://picsum.photos/id/1`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/1`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `1`,
    preview: `https://picsum.photos/id/1`,
    runTime: 100,
  },
  {
    title: `Test title 2`,
    genre: `test genre 2`,
    releaseYear: `2000`,
    thumbnail: `https://picsum.photos/id/2`,
    bgImage: `https://picsum.photos/id/2`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/2`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `1`,
    preview: `https://picsum.photos/id/2`,
    runTime: 100,
  },
];
Object.freeze(testFilms);

const testGenres = [`test genre 1`, `test genre 2`];
Object.freeze(testGenres);

export {testFilm, testFilms, testGenres};
