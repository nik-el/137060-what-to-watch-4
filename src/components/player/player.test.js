import React from 'react';
import renderer from 'react-test-renderer';
import {Player} from './player';
import {BrowserRouter as Router} from 'react-router-dom';

import {testFilm} from '../../utils/test.utils';

it(`Player renders correctly`, () => {
  const tree = renderer
    .create(
        <Router>
          <Player
            previewSrc={testFilm.preview}
          />
        </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

