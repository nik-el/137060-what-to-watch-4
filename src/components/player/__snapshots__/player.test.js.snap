// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<video
  className="small-player"
  src="https://picsum.photos/id/1"
/>
`;
